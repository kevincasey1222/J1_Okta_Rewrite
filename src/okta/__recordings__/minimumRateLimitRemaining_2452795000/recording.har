{
  "log": {
    "_recordingName": "minimumRateLimitRemaining",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.0.0"
    },
    "entries": [
      {
        "_id": "3b2c11e19d415a982b35ef7f45171a59",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "@okta/okta-sdk-nodejs/4.3.1 node/14.9.0 darwin/20.1.0"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "dev-857255-admin.okta.com"
            }
          ],
          "headersSize": 290,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://dev-857255-admin.okta.com/api/v1/users"
        },
        "response": {
          "bodySize": 1834,
          "content": {
            "mimeType": "application/json;charset=UTF-8",
            "size": 1834,
            "text": "[{\"id\":\"00u1cagurdhtxqBdf4x7\",\"status\":\"ACTIVE\",\"created\":\"2020-11-06T15:47:18.000Z\",\"activated\":\"2020-12-02T17:07:38.000Z\",\"statusChanged\":\"2020-12-02T17:18:55.000Z\",\"lastLogin\":\"2020-12-02T17:18:55.000Z\",\"lastUpdated\":\"2020-12-02T17:18:55.000Z\",\"passwordChanged\":\"2020-12-02T17:18:55.000Z\",\"type\":{\"id\":\"otys5to4jqDf0Ws0W4x6\"},\"profile\":{\"firstName\":\"Nick\",\"lastName\":\"Dowmon\",\"mobilePhone\":null,\"secondEmail\":null,\"login\":\"nick.dowmon@jupiterone.com\",\"email\":\"nick.dowmon@jupiterone.com\"},\"credentials\":{\"password\":{},\"emails\":[{\"value\":\"nick.dowmon@jupiterone.com\",\"status\":\"VERIFIED\",\"type\":\"PRIMARY\"}],\"recovery_question\":{\"question\":\"Where did you meet your spouse/significant other?\"},\"provider\":{\"type\":\"OKTA\",\"name\":\"OKTA\"}},\"_links\":{\"self\":{\"href\":\"https://dev-857255.okta.com/api/v1/users/00u1cagurdhtxqBdf4x7\"}}},{\"id\":\"00us5to77GW1mVs6w4x6\",\"status\":\"ACTIVE\",\"created\":\"2020-08-27T01:06:39.000Z\",\"activated\":null,\"statusChanged\":\"2020-08-27T01:08:57.000Z\",\"lastLogin\":\"2020-11-06T15:46:17.000Z\",\"lastUpdated\":\"2020-08-27T01:08:57.000Z\",\"passwordChanged\":\"2020-08-27T01:08:57.000Z\",\"type\":{\"id\":\"otys5to4jqDf0Ws0W4x6\"},\"profile\":{\"firstName\":\"Erkang\",\"lastName\":\"Zheng\",\"mobilePhone\":null,\"secondEmail\":null,\"login\":\"erkang.zheng@jupiterone.com\",\"email\":\"erkang.zheng@jupiterone.com\"},\"credentials\":{\"password\":{},\"emails\":[{\"value\":\"erkang.zheng@jupiterone.com\",\"status\":\"VERIFIED\",\"type\":\"PRIMARY\"}],\"recovery_question\":{\"question\":\"Where did you meet your spouse/significant other?\"},\"provider\":{\"type\":\"OKTA\",\"name\":\"OKTA\"}},\"_links\":{\"self\":{\"href\":\"https://dev-857255.okta.com/api/v1/users/00us5to77GW1mVs6w4x6\"}}},{\"id\":\"00us5ty22kvu4NAxo4x6\",\"status\":\"ACTIVE\",\"created\":\"2020-08-27T01:10:40.000Z\",\"activated\":\"2020-08-27T01:10:47.000Z\",\"statusChanged\":\"2020-08-27T11:32:46.000Z\",\"lastLogin\":\"2020-11-09T22:21:23.000Z\",\"lastUpdated\":\"2020-11-09T22:21:23.000Z\",\"passwordChanged\":\"2020-11-09T22:21:23.000Z\",\"type\":{\"id\":\"otys5to4jqDf0Ws0W4x6\"},\"profile\":{\"firstName\":\"Austin\",\"lastName\":\"Kelleher\",\"mobilePhone\":null,\"secondEmail\":null,\"login\":\"austin.kelleher@jupiterone.com\",\"email\":\"austin.kelleher@jupiterone.com\"},\"credentials\":{\"password\":{},\"emails\":[{\"value\":\"austin.kelleher@jupiterone.com\",\"status\":\"VERIFIED\",\"type\":\"PRIMARY\"}],\"recovery_question\":{\"question\":\"What was your dream job as a child?\"},\"provider\":{\"type\":\"OKTA\",\"name\":\"OKTA\"}},\"_links\":{\"self\":{\"href\":\"https://dev-857255.okta.com/api/v1/users/00us5ty22kvu4NAxo4x6\"}}},{\"id\":\"00us5z1aerBUBP41M4x6\",\"status\":\"ACTIVE\",\"created\":\"2020-08-27T01:10:14.000Z\",\"activated\":\"2020-08-27T01:10:15.000Z\",\"statusChanged\":\"2020-08-27T14:59:28.000Z\",\"lastLogin\":\"2020-08-27T14:59:28.000Z\",\"lastUpdated\":\"2020-08-27T14:59:28.000Z\",\"passwordChanged\":\"2020-08-27T14:59:28.000Z\",\"type\":{\"id\":\"otys5to4jqDf0Ws0W4x6\"},\"profile\":{\"firstName\":\"Akash\",\"lastName\":\"Ganapathi\",\"mobilePhone\":null,\"secondEmail\":null,\"login\":\"akash.ganapathi@jupiterone.com\",\"email\":\"akash.ganapathi@jupiterone.com\"},\"credentials\":{\"password\":{},\"emails\":[{\"value\":\"akash.ganapathi@jupiterone.com\",\"status\":\"VERIFIED\",\"type\":\"PRIMARY\"}],\"recovery_question\":{\"question\":\"What is the name of your first stuffed animal?\"},\"provider\":{\"type\":\"OKTA\",\"name\":\"OKTA\"}},\"_links\":{\"self\":{\"href\":\"https://dev-857255.okta.com/api/v1/users/00us5z1aerBUBP41M4x6\"}}},{\"id\":\"00us5zkpf7nGTeKXu4x6\",\"status\":\"ACTIVE\",\"created\":\"2020-08-27T01:09:43.000Z\",\"activated\":\"2020-08-27T01:09:43.000Z\",\"statusChanged\":\"2020-08-27T15:09:23.000Z\",\"lastLogin\":\"2020-12-02T17:06:34.000Z\",\"lastUpdated\":\"2020-08-27T15:09:23.000Z\",\"passwordChanged\":\"2020-08-27T15:09:23.000Z\",\"type\":{\"id\":\"otys5to4jqDf0Ws0W4x6\"},\"profile\":{\"firstName\":\"Adam\",\"lastName\":\"Williams\",\"mobilePhone\":null,\"secondEmail\":null,\"login\":\"adam.williams@jupiterone.com\",\"email\":\"adam.williams@jupiterone.com\"},\"credentials\":{\"password\":{},\"emails\":[{\"value\":\"adam.williams@jupiterone.com\",\"status\":\"VERIFIED\",\"type\":\"PRIMARY\"}],\"recovery_question\":{\"question\":\"What was your dream job as a child?\"},\"provider\":{\"type\":\"OKTA\",\"name\":\"OKTA\"}},\"_links\":{\"self\":{\"href\":\"https://dev-857255.okta.com/api/v1/users/00us5zkpf7nGTeKXu4x6\"}}}]"
          },
          "cookies": [
            {
              "expires": "1970-01-01T00:00:10.000Z",
              "name": "sid",
              "path": "/",
              "value": "[REDACTED]"
            },
            {
              "httpOnly": true,
              "name": "JSESSIONID",
              "path": "/",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Thu, 03 Dec 2020 14:54:43 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json;charset=UTF-8"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "server",
              "value": "nginx"
            },
            {
              "name": "public-key-pins-report-only",
              "value": "pin-sha256=\"r5EfzZxQVvQpKo3AgYRaT7X2bDO/kj3ACwmxfdT2zt8=\"; pin-sha256=\"MaqlcUgk2mvY/RFSGeSwBRkI+rZ6/dxe/DuQfBT/vnQ=\"; pin-sha256=\"72G5IEvDEWn+EThf3qjR7/bQSWaS2ZSLqolhnO6iyJI=\"; pin-sha256=\"rrV6CLCCvqnk89gWibYT0JO6fNQ8cCit7GGoiVTjCOg=\"; max-age=60; report-uri=\"https://okta.report-uri.com/r/default/hpkp/reportOnly\""
            },
            {
              "name": "vary",
              "value": "Accept-Encoding"
            },
            {
              "name": "x-okta-request-id",
              "value": "X8j8M0sXXXhH04n9lkevAwAAD8Y"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "p3p",
              "value": "CP=\"HONK\""
            },
            {
              "name": "x-rate-limit-limit",
              "value": "600"
            },
            {
              "name": "x-rate-limit-remaining",
              "value": "598"
            },
            {
              "name": "x-rate-limit-reset",
              "value": "1607007322"
            },
            {
              "name": "cache-control",
              "value": "no-cache, no-store"
            },
            {
              "name": "pragma",
              "value": "no-cache"
            },
            {
              "name": "expires",
              "value": "0"
            },
            {
              "name": "content-security-policy-report-only",
              "value": "default-src 'self' *.oktacdn.com dev-857255.okta.com; connect-src 'self' *.oktacdn.com *.mixpanel.com *.mapbox.com app.pendo.io data.pendo.io pendo-static-5634101834153984.storage.googleapis.com dev-857255.okta.com dev-857255-admin.okta.com https://oinmanager.okta.com data:; script-src 'unsafe-inline' 'unsafe-eval' 'self' *.oktacdn.com; style-src 'unsafe-inline' 'self' *.oktacdn.com app.pendo.io cdn.pendo.io pendo-static-5634101834153984.storage.googleapis.com; frame-src 'self' login.okta.com dev-857255.okta.com dev-857255-admin.okta.com; img-src 'self' *.oktacdn.com dev-857255.okta.com *.tiles.mapbox.com *.mapbox.com app.pendo.io data.pendo.io cdn.pendo.io pendo-static-5634101834153984.storage.googleapis.com data: blob:; font-src data: 'self' *.oktacdn.com fonts.gstatic.com; report-uri https://okta.report-uri.com/r/d/csp/reportOnly; report-to csp-report"
            },
            {
              "name": "report-to",
              "value": "{\"group\":\"csp-report\",\"max_age\":31536000,\"endpoints\":[{\"url\":\"https://okta.report-uri.com/r/d/csp/reportOnly\"}],\"include_subdomains\":true}"
            },
            {
              "name": "link",
              "value": "<https://dev-857255.okta.com/api/v1/users?limit=200>; rel=\"self\""
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=315360000; includeSubDomains"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            }
          ],
          "headersSize": 2151,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2020-12-03T14:54:42.758Z",
        "time": 303,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 303
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
