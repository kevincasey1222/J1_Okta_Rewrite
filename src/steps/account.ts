import {
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../config';

export const DATA_ACCOUNT_ENTITY = 'DATA_ACCOUNT_ENTITY';
export const ACCOUNT_ENTITY_KEY = 'okta:account';

export async function fetchAccountDetails({
  jobState,
  instance,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const accountEntity = await jobState.addEntity(
    createIntegrationEntity({
      entityData: {
        source: {
          id: `okta-account:${instance.name}`,
          name: 'Okta Account',
        },
        assign: {
          _key: `okta-account:${instance.id}`,
          _type: 'okta_account',
          _class: 'Account',
          name: `Okta - ${instance.name}`,
          displayName: `Okta - ${instance.name}`,
          webLink: instance.config.oktaOrgUrl,
        },
      },
    }),
  );

  await jobState.setData(ACCOUNT_ENTITY_KEY, accountEntity);
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-account',
    name: 'Fetch Account Details',
    entities: [
      {
        resourceName: 'Account',
        _type: 'okta_account',
        _class: 'Account',
      },
    ],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
