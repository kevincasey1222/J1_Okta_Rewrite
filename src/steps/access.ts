import * as url from 'url';

import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  parseTimePropertyValue,
  IntegrationMissingKeyError,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../client';
import { IntegrationConfig } from '../config';
import { DATA_ACCOUNT_ENTITY } from './account';
import getOktaAccountAdminUrl from '../util/getOktaAccountAdminUrl';
import { convertCredentialEmails } from '../util/convertCredentialEmails';

export const USER_GROUP_ENTITY_TYPE = 'okta_user_group';
export const APP_USER_GROUP_ENTITY_TYPE = 'okta_app_user_group';
export const MFA_DEVICE_ENTITY_TYPE = 'mfa_device';

export async function fetchUsers({
  instance,
  jobState,
  logger,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config, logger);
  const accountEntity = (await jobState.getData(DATA_ACCOUNT_ENTITY)) as Entity;

  await apiClient.iterateUsers(async (user) => {
    delete user.credentials; //no PII for you
    const webLink = url.resolve(
      getOktaAccountAdminUrl(instance.config),
      `/admin/user/profile/view/${user.id}`,
    );
    const emailProperties = convertCredentialEmails(user.credentials);
    const profile = user.profile;
    const userEntity = await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: user,
          assign: {
            _key: user.id,
            _type: 'okta_user',
            _class: 'User',
            name: `${profile.firstName} ${profile.lastName}`,
            displayName: profile.login,
            webLink: webLink,
            id: user.id,
            username: user.profile.login.split('@')[0],
            email: user.profile.email.toLowerCase(),
            verifiedEmails: emailProperties?.verifiedEmails,
            unverifiedEmails: emailProperties?.unverifiedEmails,
            status: user.status,
            active: user.status === 'ACTIVE',
            created: parseTimePropertyValue(user.created)!,
            createdOn: parseTimePropertyValue(user.created)!,
            activated: parseTimePropertyValue(user.activated)!,
            activatedOn: parseTimePropertyValue(user.activated)!,
            statusChanged: parseTimePropertyValue(user.statusChanged)!,
            statusChangedOn: parseTimePropertyValue(user.statusChanged),
            lastLogin: parseTimePropertyValue(user.lastLogin),
            lastLoginOn: parseTimePropertyValue(user.lastLogin),
            lastUpdated: parseTimePropertyValue(user.lastUpdated)!,
            lastUpdatedOn: parseTimePropertyValue(user.lastUpdated)!,
            passwordChanged: parseTimePropertyValue(user.passwordChanged),
            passwordChangedOn: parseTimePropertyValue(user.passwordChanged),
          },
        },
      }),
    );

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: userEntity,
      }),
    );

    //assign this user to their groups
    const groupIds = await apiClient.getGroupsForUser(user.id);
    for (const groupId of groupIds || []) {
      const groupEntity = await jobState.findEntity(groupId);

      if (!groupEntity) {
        throw new IntegrationMissingKeyError(
          `Expected group with key to exist (key=${groupId})`,
        );
      }

      await jobState.addRelationship(
        createDirectRelationship({
          _class: RelationshipClass.HAS,
          from: groupEntity,
          to: userEntity,
        }),
      );
    }

    //create any MFA devices assigned to this user
    if (user.status !== 'DEPROVISIONED') {
      //asking for devices for DEPROV users throws error
      const devices = await apiClient.getDevicesForUser(user.id);
      for (const device of devices || []) {
        const deviceEntity = await jobState.addEntity(
          createIntegrationEntity({
            entityData: {
              source: device,
              assign: {
                _key: device.id,
                _type: MFA_DEVICE_ENTITY_TYPE,
                _class: ['Key', 'AccessKey'],
                displayName: `${device.provider} ${device.factorType}`,
                id: device.id,
                factorType: device.factorType,
                provider: device.provider,
                vendorName: device.vendorName,
                device: device.device,
                deviceType: device.deviceType,
                status: device.status,
                created: device.created,
                lastUpdated: device.lastUpdated,
                active: device.status === 'ACTIVE',
              },
            },
          }),
        );

        await jobState.addRelationship(
          createDirectRelationship({
            _class: RelationshipClass.ASSIGNED,
            from: userEntity,
            to: deviceEntity,
          }),
        );
      }
    }
  });
}

export async function fetchGroups({
  instance,
  jobState,
  logger,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config, logger);

  const accountEntity = (await jobState.getData(DATA_ACCOUNT_ENTITY)) as Entity;

  await apiClient.iterateGroups(async (group) => {
    const webLink = url.resolve(
      getOktaAccountAdminUrl(instance.config),
      `/admin/group/${group.id}`,
    );
    const entityType =
      group.type === 'APP_GROUP'
        ? APP_USER_GROUP_ENTITY_TYPE
        : USER_GROUP_ENTITY_TYPE;

    const groupEntity = await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: group,
          assign: {
            _key: group.id,
            _type: entityType,
            _class: 'UserGroup',
            id: group.id,
            webLink: webLink,
            displayName: group.profile.name,
            created: parseTimePropertyValue(group.created)!,
            createdOn: parseTimePropertyValue(group.created)!,
            lastUpdated: parseTimePropertyValue(group.lastUpdated)!,
            lastUpdatedOn: parseTimePropertyValue(group.lastUpdated)!,
            lastMembershipUpdated: parseTimePropertyValue(
              group.lastMembershipUpdated,
            )!,
            lastMembershipUpdatedOn: parseTimePropertyValue(
              group.lastMembershipUpdated,
            )!,
            objectClass: group.objectClass,
            type: group.type,
            name: group.profile.name,
            description: group.profile.description,
          },
        },
      }),
    );

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: groupEntity,
      }),
    );
  });
}

export const accessSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-users',
    name: 'Fetch Users',
    entities: [
      {
        resourceName: 'Okta User',
        _type: 'okta_user',
        _class: 'User',
      },
      {
        resourceName: 'Okta Factor Device',
        _type: MFA_DEVICE_ENTITY_TYPE,
        _class: 'Key',
      },
    ],
    relationships: [
      {
        _type: 'okta_account_has_user',
        _class: RelationshipClass.HAS,
        sourceType: 'okta_account',
        targetType: 'okta_user',
      },
      {
        _type: 'okta_user_group_has_user',
        _class: RelationshipClass.HAS,
        sourceType: USER_GROUP_ENTITY_TYPE,
        targetType: 'okta_user',
      },
      {
        _type: 'okta_user_assigned_mfa_device',
        _class: RelationshipClass.ASSIGNED,
        sourceType: 'okta_user',
        targetType: MFA_DEVICE_ENTITY_TYPE,
      },
    ],
    dependsOn: ['fetch-groups'],
    executionHandler: fetchUsers,
  },
  {
    id: 'fetch-groups',
    name: 'Fetch Groups',
    entities: [
      {
        resourceName: 'Okta UserGroup',
        _type: USER_GROUP_ENTITY_TYPE,
        _class: 'UserGroup',
      },
      {
        resourceName: 'Okta App UserGroup',
        _type: APP_USER_GROUP_ENTITY_TYPE,
        _class: 'UserGroup',
      },
    ],
    relationships: [
      {
        _type: 'okta_account_has_user_group',
        _class: RelationshipClass.HAS,
        sourceType: 'okta_account',
        targetType: USER_GROUP_ENTITY_TYPE,
      },
      {
        _type: 'okta_account_has_app_user_group',
        _class: RelationshipClass.HAS,
        sourceType: 'okta_account',
        targetType: APP_USER_GROUP_ENTITY_TYPE,
      },
    ],
    dependsOn: ['fetch-account'],
    executionHandler: fetchGroups,
  },
];
