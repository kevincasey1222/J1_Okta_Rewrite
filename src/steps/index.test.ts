import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';

import { setupOktaRecording } from '../../test/setup/recording';
import { IntegrationConfig } from '../config';
import { fetchGroups, fetchUsers } from './access';
import { fetchAccountDetails } from './account';

const DEFAULT_ORG_URL = 'https://dev-857255.okta.com/';
const DEFAULT_API_KEY = 'dummy-api-key';

const integrationConfig: IntegrationConfig = {
  oktaOrgUrl: process.env.OKTA_ORG_URL || DEFAULT_ORG_URL,
  oktaApiKey: process.env.OKTA_API_KEY || DEFAULT_API_KEY,
};

jest.setTimeout(1000 * 60 * 1);
let recording: Recording;
afterEach(async () => {
  await recording.stop();
});

test('should collect data', async () => {
  recording = setupOktaRecording({
    directory: __dirname,
    name: 'steps',
    redactedRequestHeaders: ['Authorization'],
    redactedResponseHeaders: ['set-cookie', 'public-key-pins-report-only'],
  });

  const context = createMockStepExecutionContext<IntegrationConfig>({
    instanceConfig: integrationConfig,
  });

  // Simulates dependency graph execution.
  // See https://github.com/JupiterOne/sdk/issues/262.
  await fetchAccountDetails(context);
  await fetchGroups(context); //groups come first here
  await fetchUsers(context);

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  const accounts = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Account'),
  );
  expect(accounts.length).toBeGreaterThan(0);
  expect(accounts).toMatchGraphObjectSchema({
    _class: ['Account'],
    schema: {
      properties: {
        _type: { const: 'okta_account' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: [],
    },
  });

  const users = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('User'),
  );
  expect(users.length).toBeGreaterThan(0);
  expect(users).toMatchGraphObjectSchema({
    _class: ['User'],
    schema: {
      properties: {
        _type: { const: 'okta_user' },
        name: { type: 'string' },
        webLink: {
          type: 'string',
          format: 'url',
        },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['webLink'],
    },
  });

  const userGroups = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('UserGroup'),
  );
  expect(userGroups.length).toBeGreaterThan(0);
  expect(userGroups).toMatchGraphObjectSchema({
    _class: ['UserGroup'],
    schema: {
      properties: {
        _type: { const: 'okta_user_group' },
        name: { type: 'string' },
        webLink: {
          type: 'string',
          format: 'url',
        },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['webLink'],
    },
  });
});
