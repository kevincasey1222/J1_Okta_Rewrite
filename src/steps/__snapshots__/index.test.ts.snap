// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "okta-account:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "okta-account:Local Integration",
            "name": "Okta Account",
          },
        },
      ],
      "_type": "okta_account",
      "createdOn": undefined,
      "displayName": "Okta - Local Integration",
      "id": "okta-account:Local Integration",
      "name": "Okta - Local Integration",
      "webLink": "https://yoursubdomainhere.okta.com",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "acme-user-1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "acme-user-1",
            "name": "User One",
          },
        },
      ],
      "_type": "acme_user",
      "createdOn": undefined,
      "displayName": "User One",
      "email": "test@test.com",
      "firstName": "John",
      "id": "acme-user-1",
      "name": "User One",
      "username": "testusername",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "acme-user-2",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "acme-user-2",
            "name": "User Two",
          },
        },
      ],
      "_type": "acme_user",
      "createdOn": undefined,
      "displayName": "User Two",
      "email": "test@test.com",
      "firstName": "John",
      "id": "acme-user-2",
      "name": "User Two",
      "username": "testusername",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "acme-group-1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "acme-group-1",
            "name": "Group One",
            "users": Array [
              Object {
                "id": "acme-user-1",
              },
            ],
          },
        },
      ],
      "_type": "acme_group",
      "createdOn": undefined,
      "displayName": "Group One",
      "email": "testgroup@test.com",
      "id": "acme-group-1",
      "logoLink": "https://test.com/logo.png",
      "name": "Group One",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "okta-account:local-integration-instance",
      "_key": "okta-account:local-integration-instance|has|acme-user-1",
      "_toEntityKey": "acme-user-1",
      "_type": "okta_account_has_acme_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "okta-account:local-integration-instance",
      "_key": "okta-account:local-integration-instance|has|acme-user-2",
      "_toEntityKey": "acme-user-2",
      "_type": "okta_account_has_acme_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "okta-account:local-integration-instance",
      "_key": "okta-account:local-integration-instance|has|acme-group-1",
      "_toEntityKey": "acme-group-1",
      "_type": "okta_account_has_acme_group",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "acme-group-1",
      "_key": "acme-group-1|has|acme-user-1",
      "_toEntityKey": "acme-user-1",
      "_type": "acme_group_has_user",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "okta_account",
    "acme_user",
    "okta_account_has_acme_user",
    "acme_group",
    "okta_account_has_acme_group",
    "acme_group_has_user",
  ],
  "numCollectedEntities": 4,
  "numCollectedRelationships": 4,
}
`;
